src/main/java/com/rabbitmq/http/client/ReactorNettyClient.java

```diff
@@ -535,6 +541,98 @@ public static String basicAuthentication(String username, String password) {
         return doDelete("parameters", "shovel", enc(vhost), enc(shovelName));
     }
 
+    //
+    // Federation support
+    //
+    /**
+     * Declares an upstream
+     * @param vhost virtual host for which to declare the upstream
+     * @param name name of the upstream to declare
+     * @param details upstream arguments
+     */
+    public Mono<HttpResponse> declareUpstream(String vhost, String name, UpstreamDetails details) {
+        if (StringUtils.isEmpty(details.getUri())) {
+            throw new IllegalArgumentException("Upstream uri must not be null or empty");
+        }
+        ParameterWrapper<UpstreamDetails> body = new ParameterWrapper<UpstreamDetails>()
+                .setComponent("federation-upstream")
+                .setVhost(vhost)
+                .setName(name)
+                .setValue(details);
+        return doPut(body, "parameters", "federation-upstream", enc(vhost), enc(name));
+    }
+
+    /**
+     * Deletes an upstream
+     * @param vhost virtual host for which to delete the upstream
+     * @param name name of the upstream to delete
+     */
+    public Mono<HttpResponse> deleteUpstream(String vhost, String name) {
+        return doDelete("parameters", "federation-upstream", enc(vhost), enc(name));
+    }
+
+    /**
+     * Returns a list of upstreams for "/" virtual host
+     */
+    public Flux<UpstreamInfo> getUpstreams() {
```

Comment:
Why using `Flux<UpstreamInfo>` and not `Flux<ParameterWrapper<UpstreamDetails>>`? Or the other way around? Both clients should be consistent.

Reply:
It seems simpler and more coherent with the other API methods to return *Info types without generics. So I changed the synchronous client in the new commit to return *Info types. Is it fine, or is it better to make both clients return generalized RuntimeParameter?

To return `Flux<RuntimeParameter<UpstreamDetails>>`, can do the following - overload doGetMono, decode, deserialize and doGetFlux with copies that accept type data for ObjectMapper (JavaType and TypeReference are from Jackson lib).
```
    /**
     * Returns a list of upstreams for "/" virtual host
     */
    public Flux<RuntimeParameter<UpstreamDetails>> getUpstreams() {
        TypeReference<RuntimeParameter<UpstreamDetails>> type =
                new TypeReference<RuntimeParameter<UpstreamDetails>>() {};
        return doGetFlux(type, "parameters", "federation-upstream");
    }

    /**
     * Returns a list of upstream sets for "/" virtual host
     */
    public Flux<RuntimeParameter<List<UpstreamSetDetails>>> getUpstreamSets() {
        TypeReference<RuntimeParameter<List<UpstreamSetDetails>>> type =
                new TypeReference<RuntimeParameter<List<UpstreamSetDetails>>>() {};
        return doGetFlux(type, "parameters", "federation-upstream-set");
    }


    private <T> Mono<T> doGetMono(JavaType type, String... pathSegments) {
        return Mono.from(client
                .headersWhen(authorizedHeader())
                .get()
                .uri(uri(pathSegments))
                .response(decode(type)));
    }

    protected <T> BiFunction<? super HttpClientResponse, ? super ByteBufFlux, Publisher<T>> decode(JavaType type) {
        return (response, byteBufFlux) -> {
            this.responseCallback.accept(response);
            if (response.status().code() == 404) {
                return Mono.empty();
            } else {
                return byteBufFlux.aggregate().asInputStream().map(bytes -> deserialize(bytes, type));
            }
        };
    }

    private <T> T deserialize(InputStream inputStream, JavaType type) {
        try {
            T value = objectMapper.readValue(inputStream, type);
            inputStream.close();
            return value;
        } catch (IOException e) {
            throw Exceptions.propagate(e);
        }
    }

    @SuppressWarnings("unchecked")
    private <T> Flux<T> doGetFlux(TypeReference<T> type, String... pathSegments) {
        ArrayType arrayType = objectMapper.getTypeFactory().constructArrayType(objectMapper.getTypeFactory().constructType(type));
        return (Flux<T>) doGetMono(arrayType, pathSegments).flatMapMany(items -> Flux.fromArray((Object[]) items));
    }
```